### DISCUSSION TITLE
E2Table Class in API

### DISCUSSION THREAD

**Discussion by Denneisk** [wiremod/wire#2817] [Suggestions]
Make E2Tables a class. This would be pretty useful as parts of the implementation and functionality of E2 tables could be offloaded into E2Lib (or wherever the class is defined), thus making them easier to operate with.

Provisional API:
- `class E2Table`
  - `integer size`
  - `table<integer, any> n`
  - `table<integer, string> ntypes`
  - `table<string, any> s`
  - `table<string, string> stypes`
  - `string? IsHomogeneous` - ? Optimization flag. Set to the homogeneous typeid
  - `boolean? IsSequential` - ? Optimization flag
- `newE2Table(table? args, table? types): E2Table` - Initializes an `E2Table`. If first argument and second argument are specified, initialize the output `E2Table` with the key-values in `args` and the types in `types`. The second argument may be optional if type guessing is included.
- `E2Table:Get(string|integer key): any? value, string? type` - Returns the value and type of an element at the index or nil if it doesn't exist.
- `E2Table:Set(string|integer key, any value, string type)` - Sets the key in the `E2Table` to the value with the provided type. The third argument may be optional if type guessing is included.
- `E2Table:Clone(): E2Table` - Shallow copy (subtables are references) (Provisional name)
- `E2Table:Duplicate(): E2Table` - ? Deep copy (subtables are new tables) (Provisional name)

```lua
e2function void myFunc(table t)
    t:Set(1, "Test", "s")

    local val, typeid = t:Get(1)
    if typeid == "s" then print(val) end -- prints "Test"

    val = t:GetOfType("1", "s") -- ? Returns nil if type mismatch
    if val then print(val) end -- Does not execute

    val = t:GetOfTypeOrZero(1, "n") -- ?? Returns "zero" of desired type if type mismatch
    print(val) -- prints "0"

    foreach k, v in t:IterN("s") do
        -- ????? Typed iterator
    end
end
```

Just spitballing with these last ones, but kind of a must if you want to abstract E2 tables' implementations from their usage.

Could also include a type guessing step here and there but that *demands* a better type implementation.

**Comment by vurvdev** [comment]
I like the idea, tables should've been like this in the first place for sure. E2 exposes too many internal behaviors that lock us into these backwards compatibility situations.

API might need to be a bit different, I alongside others would definitely appreciate arbitrary key types, although maybe that could be done with a separate type since adjusting this might be too annoying while balancing backwards compatibility.

**Reply by Denneisk** [reply]
~~Arbitrary type is definitely desirable but I simply don't trust types in E2 to not accidentally mismatch and I'd rather not use type guessing.~~ Never mind misread this

**Reply by Denneisk** [reply]
Also, in my opinion, it only needs to act as a helper for not needing to write your own table initializer, so existing table access should be fine.
